### DISCUSSION TITLE
Custom structure as InputArgument of method

### DISCUSSION THREAD

**Discussion by AndreyKl4** [FreeOpcUa/opcua-asyncio#448] [Q&A]
Hi all!

I have faced such problem: I created a custom data type with help data type definitions. After that I used custom data type as InputArgument of method. I found out that it works only in case when I set up argument as an array:
arg.ValueRank = 0
arg.ArrayDimensions = []. 
But, actually, i need to use my Argument as Scalar with ValueRank = -1. When i set up ValueRank = -1 it displays incorrectly.
Perhaps I understand smth incorrect, when we use data type  definitions we get an array of fields and that is why i have to use ValueRank = 0 instead of ValueRank = -1 ? 

Thanks in Advance!

Andrii Klimov

**Comment by AndreasHeine** [comment]
have you seen this? https://github.com/FreeOpcUa/opcua-asyncio/blob/master/examples/server-extension-object-as-method-argument.py

**Comment by AndreyKl4** [comment]
@AndreasHeine yes i have. It was first what i looked at. In my case it dispalys incorrectly ...

**Reply by AndreasHeine** [reply]
could you be more precise? mabe with a screenshots so i dont get it wrong!

**Reply by AndreyKl4** [reply]
Argument : 
```python
 async def create_arg_from_structure(self, struct_type, struct_name):
        arg = ua.Argument()
        arg.Name = struct_name
        arg.DataType = struct_type[0].nodeid
        arg.ValueRank = 0 # -1
        arg.ArrayDimensions = []
        return arg
```
![valuerank_0](https://user-images.githubusercontent.com/73280686/108047645-54f07b00-7046-11eb-8a4e-3e612633c485.png)
![valuerank_-1](https://user-images.githubusercontent.com/73280686/108047647-55891180-7046-11eb-8421-7c81a10b14f7.png)

Thanks for your help!

**Reply by AndreyKl4** [reply]
In the first case i use arg.ValueRank = 0 in the second arg.ValueRank = -1 with the same data type.

**Reply by AndreyKl4** [reply]
it is datatype: 
![datatype](https://user-images.githubusercontent.com/73280686/108048147-f8419000-7046-11eb-87c3-2e8e373cc2ef.png)

**Reply by AndreasHeine** [reply]
you wrote "...displays incorrectly."

could you tell me what you try to setup? MethodInArgs? MethodOutArgs? What is actually wrong or appears wrong?

**Reply by AndreyKl4** [reply]
I tried to set up MethodInArg in this case. Idea was to create a custom datatype(see screenshot 1 and 3) and use it as MethodInArg as Scalar(ValueRanks = -1) but not as an Array(ValueRank = 0) of such datatypes...  When I try to set up it as Scalar a I get (see screenshot 2) but i want to get(see Screenshot 1) but not an Array... Perhaps, I do smth wrong...

**Reply by AndreasHeine** [reply]
MethodInArg is always an Array of Args even if it is only one

https://reference.opcfoundation.org/v104/Core/docs/Amendment3/5.7.3/

> inputArguments [] | BaseDataType | List of input argument values. An empty list indicates that  there are no input arguments. The size and order of this list matches  the size and order of the input arguments defined by the input InputArguments Property of the Method.The name, a description and the data type of each argument are defined by the Argument structure in each element of the method’s InputArguments Property.If a Method has optional input arguments, these optional input arguments are indicated by the Method’s DefaultInputValues Property. In this case fewer arguments than the total number of input arguments defined may be passed by the Client. The InputArguments and DefaultInputValues Property are defined in Part 3.
> -- | -- | --

**Reply by AndreasHeine** [reply]
if you cant share the sourcecode send me an email!

**Reply by AndreyKl4** [reply]
@AndreasHeine thanks for your help. I have send u an email
